kratos/processes/reorder_and_optimize_modelpart_process.cpp

```diff
@@ -37,71 +38,59 @@ namespace Kratos
-            }
+        IndexPartition<std::size_t>(mrModelPart.Conditions().size()).for_each([&](std::size_t Index){
+            (mrModelPart.ConditionsBegin() + Index)->SetId(Index+1);
+        });
+        mrModelPart.Conditions().Sort();
 
-            #pragma omp parallel for
-            for(int i=0; i<static_cast<int>(mrModelPart.Elements().size()); ++i)
-            {
-                Element::Pointer& pelem = *(mrModelPart.Elements().ptr_begin() + i);
+        OptimizeOrdering();
 
-                Geometry<Node<3>>::PointsArrayType tmp;
-                const auto& geom = pelem->GetGeometry();
-                tmp.reserve(geom.size());
-                for(unsigned int k=0; k<geom.size(); ++k)
-                    tmp.push_back( mrModelPart.Nodes()(geom[k].Id()) );
+        //make a parallel clone of all the nodes
+        IndexPartition<std::size_t>(mrModelPart.Nodes().size()).for_each([&](std::size_t Index){
+            Node<3>::Pointer& pnode = *(mrModelPart.Nodes().ptr_begin() + Index);
+            pnode = pnode->Clone();
+        });
 
-                auto paux = pelem->Create(pelem->Id(), tmp, pelem->pGetProperties());
+        block_for_each(mrModelPart.Elements(), [&](Element& rElem){
+            Geometry<Node<3>>::PointsArrayType tmp;
+            const auto& geom = rElem.GetGeometry();
+            tmp.reserve(geom.size());
+            for(unsigned int k=0; k<geom.size(); ++k)
+                tmp.push_back( mrModelPart.Nodes()(geom[k].Id()) );
 
-                paux->Data() = pelem->Data();
+            auto paux = rElem.Create(rElem.Id(), tmp, rElem.pGetProperties());
 
-                pelem = paux;
-            }
+            paux->Data() = rElem.Data();
 
-            #pragma omp parallel for
-            for(int i=0; i<static_cast<int>(mrModelPart.Conditions().size()); ++i)
-            {
-                Condition::Pointer& pcond = *(mrModelPart.Conditions().ptr_begin() + i);
+            rElem = (*paux);
+        });
 
-                Geometry<Node<3>>::PointsArrayType tmp;
-                const auto& geom = pcond->GetGeometry();
-                tmp.reserve(geom.size());
-                for(unsigned int k=0; k<geom.size(); ++k)
-                    tmp.push_back( mrModelPart.Nodes()(geom[k].Id()));
+        block_for_each(mrModelPart.Conditions(), [&](Condition& rCond){
+            Geometry<Node<3>>::PointsArrayType tmp;
+            const auto& geom = rCond.GetGeometry();
+            tmp.reserve(geom.size());
+            for(unsigned int k=0; k<geom.size(); ++k)
+                tmp.push_back( mrModelPart.Nodes()(geom[k].Id()));
 
-                auto paux = pcond->Create(pcond->Id(), tmp, pcond->pGetProperties());
+            auto paux = rCond.Create(rCond.Id(), tmp, rCond.pGetProperties());
 
-                paux->Data() = pcond->Data();
+            paux->Data() = rCond.Data();
 
-                pcond = paux;
-            }
+            rCond = (*paux);
+        });
```

Comment:
I don't think you can change this
Better stick with the pointers and use `IndexPartition`